src/snowflake/snowpark/mock/plan.py

```diff
@@ -895,6 +909,215 @@ def calculate_expression(
+
+        res_index = res.index  # List of row indexes of the result
+
+        # Process partition_by clause
+        if window_spec.partition_spec:
+            res = res.groupby(
+                [exp.name for exp in window_spec.partition_spec],
+                sort=False,
+                as_index=False,
+            )
+            res_index = []
+            for r in res:
+                res_index += list(r[1].index)
+
+        if not window_spec.frame_spec or not isinstance(
+            window_spec.frame_spec, SpecifiedWindowFrame
+        ):
+            # TODO: If no window frame is specified, rank related function uses the entire window, other functions use cumulative window
+            indexer = EntireWindowIndexer()
+            res = res.rolling(indexer)
+            windows = [input_data.loc[w.index] for w in res]
+
+        elif isinstance(window_spec.frame_spec.frame_type, RowFrame):
+            indexer = RowFrameIndexer(frame_spec=window_spec.frame_spec)
+            res = res.rolling(indexer)
+            windows = [w for w in res]
+
+        elif isinstance(window_spec.frame_spec.frame_type, RangeFrame):
+            assert window_spec.order_spec and len(window_spec.order_spec) == 1
+            order_by_val = window_spec.order_spec[0].child
+
+            _indexer = EntireWindowIndexer()
+            _res = res.rolling(_indexer)
+
+            windows = []
+            upper = window_spec.frame_spec.upper
+            lower = window_spec.frame_spec.lower
+
+            for current_row, window in zip(res_index, _res):
+                row_idx = list(window.index).index(
+                    current_row
+                )  # the row's 0-base index in the window
+                order_by_values = calculate_expression(
+                    order_by_val, window, analyzer, expr_to_alias, keep_literal=True
+                ).values
+                # bisect
+                if isinstance(lower, CurrentRow):
+                    left_idx = row_idx
+                elif isinstance(lower, UnboundedPreceding):
+                    left_idx = 0
+                elif isinstance(lower, Literal):
+                    lower_bound = order_by_values.loc[current_row] - lower.value
+                    left_idx = min(
+                        bisect_left(order_by_values, lower_bound), row_idx
+                    )  # bisect_left might return len(order_values)
+
+                if isinstance(upper, CurrentRow):
+                    right_idx = row_idx
+                elif isinstance(upper, UnboundedFollowing):
+                    right_idx = len(window) - 1
+                elif isinstance(upper, Literal):
+                    upper_bound = order_by_values.loc[current_row] + upper.value
+                    right_idx = max(bisect_right(order_by_values, upper_bound), row_idx)
+
+                windows.append(window.iloc[[i for i in range(left_idx, right_idx + 1)]])
+
+        assert len(windows) == len(input_data)
```

Comment:
instead of assertion error, I think we can raise a more user friend error if the size doesn't match?

Reply:
This is more of a sanity check instead of user facing error, I'll remove it.